File: java/ExampleProfile.java

class BailOut extends Exception {}

class ExampleProfile {
	private static final int BOARD_SIZE = 8;

	private static void validate(int[] queens) throws BailOut {
		int left, right, col;
		left = right = col = queens[queens.length-1];
	    for(int n = queens.length-2; n>=0; n--) {
			int r = queens[n];
	        left--;
			right++;
	        if(r == left || r == col || r == right) {
	            throw new BailOut();
			}
		}
	}

	private static int[] add_queen(int[] queens) throws BailOut {
		try {Thread.sleep(10);}
		catch(Exception e) {}

		for(int i=0; i<BOARD_SIZE; i++) {
			int[] test_queens = new int[queens.length + 1];
			for(int n=0; n<queens.length; n++) {
				test_queens[n] = queens[n];
			}
			test_queens[queens.length] = i;

			try {
				validate(test_queens);
				if(test_queens.length == BOARD_SIZE) {
					return test_queens;
				}
				else {
					return add_queen(test_queens);
				}
			}
			catch(BailOut e) {}
		}
		throw new BailOut();
	}

	public static void main(String[] args) {
		int[] queens = new int[0];
		try {
			queens = add_queen(queens);
			for(int n=0; n<queens.length; n++) {
				System.out.print(queens[n]);
			}
			System.out.println("");
		}
		catch(BailOut e) {
			System.out.println("Top-level bail");
		}
//		for(int n=0; n<BOARD_SIZE; n++) {
//			int q = queens[n];
//			System.out.println(". "*q + "Q " + ". "*(BOARD_SIZE - q - 1));
//		}
	}
}

File: python/example-profile.py
#!/usr/bin/python

import context.api as c
import time

c.set_log("file://output.py-profile.ctxt")
c.set_profile(True)

BOARD_SIZE = 8

class BailOut(Exception):
    pass

def validate(queens):
    time.sleep(0.01)

    left = right = col = queens[-1]
    for r in reversed(queens[:-1]):
        left, right = left - 1, right + 1
        if r in (left, col, right):
            raise BailOut

def add_queen(queens):
    for i in range(BOARD_SIZE):
        test_queens = queens + [i]
        try:
            validate(test_queens)
            if len(test_queens) == BOARD_SIZE:
                return test_queens
            else:
                return add_queen(test_queens)
        except BailOut:
            pass
    raise BailOut

queens = add_queen([])
print queens
print "\n".join(". "*q + "Q " + ". "*(BOARD_SIZE - q - 1) for q in queens)
